cclib/method/ddec.py

```diff
@@ -590,9 +663,256 @@ def condition_densities(self):
+            N_A.append(self._rho_A[atomi].integrate())
+            self.rho_wavg.append(
+                (self._theta[atomi] + self._rho_A_avg[atomi] * self._wavg[atomi] / 5.0)
+                / (1.0 - (4.0 / 5.0) * self._wavg[atomi])
+            )
+            u_A.append(self._integrate_from_radial([self._theta[atomi]], [atomi]))
+        self.u_A.append(u_A)
+
+        return N_A
+
+    def reshape_G(self):
+        """ Calculate G_A(r_A) and reshape densities
+            [STEP 4-7]
+            TODO: some similarities exist with step 3 routine. code refactoring may improve readability.
         """
+        self._candidates_bigPhi = []
+        self._candidates_phi = []
+
+        # Initial conditions are detailed in Figure S3 in doi: 10.1039/c6ra04656h
+        phiAII = numpy.zeros_like(self.data.atomnos, dtype=float)
+        bigphiAII = numpy.zeros_like(self.data.atomnos, dtype=float)
+        self._g = []
+
+        for atomi in range(self.data.natom):
+            # G_A -- equation S102 in doi: 10.1039/c6ra04656h
+            self._g.append(
+                numpy.zeros_like(self.proatom_density[atomi], dtype=float)
+            )
+            self._g[atomi] = self.rho_wavg[atomi] + bigphiAII[atomi] * numpy.sqrt(self.rho_wavg[atomi])
+            # Exponential constraint (as expressed in equation S105)
+            eta = (1-(self.tau[atomi])**2) * 1.75 * convertor(1, "Angstrom", "bohr")
+            exp_applied = self._g[atomi][:-1] * numpy.exp(-1 * eta[1:] * numpy.diff(self.radial_grid_r[atomi]))
+            for radiusi in range(1, len(self._g[atomi])):
+                self._g[atomi][radiusi] = min(self._g[atomi][radiusi], exp_applied[radiusi - 1])
+            # phi_A^II -- Equation S106 in doi: 10.1039/c6ra04656h
+            phiAII[atomi] = self._integrate_from_radial([self._g[atomi] - self.rho_wavg[atomi]], [atomi])
+
+            self._candidates_bigPhi.append([bigphiAII[atomi]])
+            self._candidates_phi.append([phiAII[atomi]])
+
+            fitphi = True  # when convergence is reached, this is modified to False.
+
+            while phiAII[atomi] <= 0:
+                # Iterative algorithm until convergence
+                # Refer to Figure S3 in doi: 10.1039/c6ra04656h
+                bigphiAII[atomi] = 2 * bigphiAII[atomi] - phiAII[atomi] / self._integrate_from_radial(
+                    [numpy.sqrt(self.rho_wavg[atomi])], [atomi]
+                )
+
+                # When Phi is updated, related quantities are updated as well
+                # Refer to S102 in doi: 10.1039/c6ra04656h
+                phiAII[atomi], self._g[atomi] = self._phiaii(
+                    self.rho_wavg[atomi], eta, bigphiAII[atomi], atomi
+                )
+
+                self._candidates_phi[atomi].append(phiAII[atomi])
+                self._candidates_bigPhi[atomi].append(bigphiAII[atomi])
+
+            # lowerbigPhi is largest negative Phi.
+            # upperbigPhi is smallest positive Phi.
```

Comment:
Don't you mean largest positive and smallest negative? It sounds like these are bounds.

Reply:
These are largest negative and smallest positive, because the algorithm tries to find two closest numbers to zero. (details are in Figure S1) Let me know if I missed something. Thanks.

Reply:
Something with "the smallest magnitude negative and positive phi" would be less confusing.